### PR TITLE
feat: Enable sorting and finish pagination on Table Queries

### PR DESCRIPTION
**Pull request by tssweeney**
This PR adds the `sort_by` field to our `table_query` and finishes the pagination implementation. Time to make Table views fast!

### CODE REVIEW

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 1]
```diff
@@ -767,19 +769,39 @@ def add_new_row_needed_to_insert(row_data: Any) -> str:
 
     def table_query(self, req: tsi.TableQueryReq) -> tsi.TableQueryRes:
         conds = []
-        parameters = {}
+        parameters: Dict[str, Any] = {}
         if req.filter:
             if req.filter.row_digests:
-                conds.append("tr.digest IN {row_digets: Array(String)}")
+                conds.append("tr.digest IN {row_digests: Array(String)}")
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:775]: yay for testing! Found a bug unrelated to this PR! (:

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 2]
```diff
@@ -767,19 +769,39 @@ def add_new_row_needed_to_insert(row_data: Any) -> str:
 
     def table_query(self, req: tsi.TableQueryReq) -> tsi.TableQueryRes:
         conds = []
-        parameters = {}
+        parameters: Dict[str, Any] = {}
         if req.filter:
             if req.filter.row_digests:
-                conds.append("tr.digest IN {row_digets: Array(String)}")
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:773]: !

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 3]
```diff
@@ -767,19 +769,39 @@ def add_new_row_needed_to_insert(row_data: Any) -> str:
...
                 parameters["row_digests"] = req.filter.row_digests
         else:
             conds.append("1 = 1")
+
+        sort_clause = ""
+        pb = ParamBuilder()
+        if req.sort_by:
+            sort_fields = []
+            for i, sort in enumerate(req.sort_by):
+                # TODO: better splitting of escaped dots (.) in field names
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:785]: 😢

File: weave/trace_server/trace_server_interface.py [hunk 4]
```diff
@@ -547,15 +547,38 @@ class TableCreateRes(BaseModel):
 
 
 class TableRowFilter(BaseModel):
-    row_digests: Optional[List[str]] = None
+    row_digests: Optional[List[str]] = Field(
```
Comment (gtarpenning) [inline weave/trace_server/trace_server_interface.py:550]: might be nice to show an example of what a digest actually looks like: 

`aonareimsvtl13apimtalpa4435rpmgnaemrpgmarltarstaorsnte134avrims`

File: weave/trace_server/trace_server_interface.py [hunk 5]
```diff
@@ -547,15 +547,38 @@ class TableCreateRes(BaseModel):
...
+        default=None, description="Maximum number of rows to return", examples=[100]
+    )
+    offset: Optional[int] = Field(
+        default=None,
+        description="Number of rows to skip before starting to return rows",
+        examples=[10],
+    )
+    sort_by: Optional[List[SortBy]] = Field(
+        default=None,
+        description="List of fields to sort by. Fields can be dot-separated to access dictionary values. No sorting uses the default table order.",
```
Comment (gtarpenning) [inline weave/trace_server/trace_server_interface.py:579]: what is the default table order

File: weave/trace_server/sqlite_trace_server.py [hunk 6]
```diff
@@ -810,13 +813,77 @@ def add_new_row_needed_to_insert(row_data: Any) -> str:
...
+            JOIN table_rows tr ON OrderedDigests.digest = tr.digest
+        WHERE {predicate}
+        {order_by}
+        """
+
+        if req.limit is not None:
+            query += f" LIMIT {req.limit}"
+        if req.offset is not None:
+            if req.limit is None:
+                query += " LIMIT -1"
```
Comment (gtarpenning) [inline weave/trace_server/sqlite_trace_server.py:874]: is that required by offset? interesting

File: weave/tests/trace/test_table_query.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    result_digests = [r.digest for r in res.rows]
+
+    sorted_data = sorted(data, key=lambda x: x["id"], reverse=True)
+    expected_data = sorted_data[offset : offset + limit]
+
+    assert len(res.rows) == limit
+    assert result_vals == expected_data
+
+
+def test_table_query_multiple_sort_criteria(client: WeaveClient):
```
Comment (gtarpenning) [inline weave/tests/trace/test_table_query.py:174]: might also want to add a test for paging/offset with a query that has a non-existent digest/table.

**Review by gtarpenning — APPROVED** [review]
Nice. Just some micro questions as usual

**Pull request merged**